libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp

```diff
@@ -998,14 +999,373 @@ bool AP_Frsky_Telem::get_telem_data(uint8_t &frame, uint16_t &appid, uint32_t &d
             singleton->init();
         }
     }
-    if (!singleton) {
+    if (singleton == nullptr) {
         return false;
     }
     return singleton->_get_telem_data(frame, appid, data);
 }
 
-namespace AP {
-    AP_Frsky_Telem *frsky_telem() {
-        return AP_Frsky_Telem::get_singleton();
+#if HAL_WITH_FRSKY_TELEM_BIDIRECTIONAL
+/*
+  allow external transports (e.g. FPort), to supply telemetry data
+ */
+bool AP_Frsky_Telem::_set_telem_data(const uint8_t frame, const uint16_t appid, const uint32_t data)
+{
+    // queue only Uplink packets
+    if (frame == SPORT_UPLINK_FRAME || frame == SPORT_UPLINK_FRAME_RW) {
+        const FRSky::sport_packet_t sp {
+            0x0D,
```

Comment:
`SENSOR_ID_UPLINK` ?

That might require putting a type in instead of just having a define.

Reply:
that byte is ignored, I'll put 0x00 and add a comment